crates/pipeline_manager/src/pipeline_automata.rs

```diff
@@ -377,55 +377,49 @@ impl<T: PipelineExecutor> PipelineAutomaton<T> {
                             } else {
-                                self.force_kill_pipeline_on_error(&mut pipeline, status, &body)
+                                self.mark_pipeline_as_failed_on_error(&mut pipeline, status, &body)
                                     .await?;
                             }
                         }
                     }
                 }
-                // Shutdown request.  Forward to the pipeline; on success go to the `ShuttingDown`
-                // state.
+                // Issue a pipeline shutdown.
                 (PipelineStatus::Running, PipelineStatus::Shutdown)
                 | (PipelineStatus::Paused, PipelineStatus::Shutdown) => {
-                    // TODO: replace with direct call to shutdown
-                    match pipeline_http_request_json_response(
-                        self.pipeline_id,
-                        Method::GET,
-                        "shutdown",
-                        &pipeline.location,
-                    )
-                    .await
-                    {
-                        Err(e) => {
-                            self.force_kill_pipeline(&mut pipeline, Some(e)).await?;
+                    match self.pipeline_handle.shutdown().await {
+                        Ok(_) => {
+                            self.update_pipeline_status(
+                                &mut pipeline,
+                                PipelineStatus::ShuttingDown,
+                                None,
+                            )
+                            .await;
+                            self.update_pipeline_runtime_state(&pipeline).await?;
+                            poll_timeout = Self::SHUTDOWN_POLL_PERIOD;
                         }
-                        Ok((status, body)) => {
-                            if status.is_success() {
-                                self.update_pipeline_status(
-                                    &mut pipeline,
-                                    PipelineStatus::ShuttingDown,
-                                    None,
-                                )
-                                .await;
-                                self.update_pipeline_runtime_state(&pipeline).await?;
-                                poll_timeout = Self::SHUTDOWN_POLL_PERIOD;
-                            } else {
-                                self.force_kill_pipeline_on_error(&mut pipeline, status, &body)
-                                    .await?;
-                            }
+                        Err(e) => {
+                            self.mark_pipeline_as_failed(
+                                &mut pipeline,
+                                Some(RunnerError::PipelineShutdownError {
+                                    pipeline_id: self.pipeline_id,
+                                    error: format!(
+                                        "pipeline_handle.shutdown() returned an error: {e}"
```

Comment:
This is probably not a great user-visible error message. Can this be just `e.to_string()?`